### PR TITLE
[ORCA] Fix unused nested CTE pruning 

### PR DESCRIPTION
**Pull request by dgkimura**
In the case of nested CTE, additional care needs to be taken to remove
unused anchors that have no corresponding CTE consumer. Consider the
following query:

```sql
EXPLAIN WITH q AS (
  SELECT * FROM (WITH cte AS (SELECT 1) SELECT 1) t)
SELECT * FROM q;
```

Prior to this commit, it produced two common table expressions:

1.
        +--CLogicalCTEProducer (1), Columns: ["?column?" (1)]
           +--CLogicalProject
              |--CLogicalConstTableGet Columns: [(0)] Values: [(1)]
              +--CScalarProjectList
                 +--CScalarProjectElement "?column?" (1)
                    +--CScalarConst (1)

2.
        +--CLogicalCTEProducer (0), Columns: ["?column?" (3)]
           +--CLogicalCTEAnchor (1)
              +--CLogicalProject
                 |--CLogicalConstTableGet Columns: [(2)] Values: [(1)]
                 +--CScalarProjectList
                    +--CScalarProjectElement "?column?" (3)
                       +--CScalarConst (1)

However, `CLogicalCTEAnchor (1)` should be pruned because there is no
corresponding consumer. That leads to an ORCA fallback:

    ERROR,"CXformCTEAnchor2TrivialSelect.cpp:56: Failed assertion: 0 < ulConsumers

This commit addresses the issue by marking unused CTEs immediately
before pruning unused CTEs.

### CODE REVIEW

**Review by gpopt — COMMENTED** [review]
Can we have a mdp test?

File: src/test/regress/expected/qp_with_clause.out [hunk 1]
```diff
@@ -11111,6 +11111,24 @@ SELECT * FROM c as c1, zoo WHERE zoo.c = c1.b;
      3 | 4 | 4 | 4
 (1 row)
 
+-- Test that nested CTE generates a valid plan
+EXPLAIN WITH q AS (SELECT * FROM (WITH cte AS (SELECT * FROM car) SELECT * FROM car WHERE a > 7) t) SELECT * FROM q;
+                                             QUERY PLAN                                             
+----------------------------------------------------------------------------------------------------
+ Gather Motion 3:1  (slice1; segments: 3)  (cost=10000000000.00..10000000871.08 rows=28700 width=8)
```
Comment (gpopt) [inline src/test/regress/expected/qp_with_clause.out:11118]: Can we remove cost info?

**Comment by DevChattopadhyay** [comment]
Just for my better understanding on CTE's.. For the query mentioned in the commit message why we have two algebrized trees. Is it because we have two CTE definitions in the query. What is the significance of CLogicalCTEAnchor and how we can say that there is no corresponding consumer for it by looking at the algebrized query

**Review by gpopt — APPROVED** [review]
+1

**Comment by dgkimura** [comment]
On Tuesday, September 26, 2023 6:13 AM Dev Swaroop Chattopadhyay wrote:
> For the query mentioned in the commit message why we have two algebrized trees. Is it because we have two CTE definitions in the query.

Yes.

> What is the significance of CLogicalCTEAnchor and how we can say that there is no corresponding consumer for it by looking at the algebrized query

A CTE plan needs a way to ensure the producer is run before the consumer.  An anchor is like a placeholder so that we can add the producer into the needed places in the plan.  In the example plan, there was a `CLogicalCTEAnchor` without a corresponding `CPhysicalCTEConsumer` underneath.  `CExpressionPreprocessor::PexprRemoveUnusedCTEs` is supposed to eliminate those cases, but the example was a case where it did not.

File: src/test/regress/sql/qp_with_clause.sql [hunk 2]
```diff
@@ -10346,6 +10346,10 @@ SELECT * FROM c as c1, zoo WHERE zoo.c != 4 AND c1.b = zoo.c
 UNION ALL
 SELECT * FROM c as c1, zoo WHERE zoo.c = c1.b;
 
+-- Test that nested CTE generates a valid plan
+EXPLAIN (COSTS OFF) WITH q AS (SELECT * FROM (WITH cte AS (SELECT * FROM car) SELECT * FROM car WHERE a > 7) t) SELECT * FROM q;
+WITH q AS (SELECT * FROM (WITH cte AS (SELECT * FROM car) SELECT * FROM car WHERE a > 7) t) SELECT * FROM q;
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/test/regress/sql/qp_with_clause.sql:10351]: Thanks for trying it out.  Yeah, that matches my understanding.

**Review by DevChattopadhyay — APPROVED** [review]
Thanks @dgkimura for the explanation. The changes look good to me.

**Review by 24nishant — APPROVED** [review]
Thanks for the details @dgkimura , the changes looks good to me.

**Comment by dgkimura** [comment]
Merged. Thanks for the reviews!

**Pull request merged**